File: python_files/raylib/flappy_bird.py
import pyray as p
import random


SCREEN_WIDTH = 400
SCREEN_HEIGHT = 600
SCREEN_TITLE = "Flappy Bird"
SCREEN_BACKGROUND = p.SKYBLUE
GAME_FPS = 60


def main():
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)
    p.set_target_fps(GAME_FPS)

    score = 0
    game_over = False

    rect = p.Rectangle(20, 20, 30, 30)
    rect_color = p.RED
    gravity = 1
    jump_force = -12
    change_y = 0

    vertical_gap = 150
    horitzontal_gap = 250

    pillar_x = p.get_screen_width()
    pillar_y = 0
    pillar_width = 50

    pillar_speed = 2
    pillar_color = p.DARKGREEN

    pillar_list = []

    while not p.window_should_close():
        if not game_over:
            if (
                len(pillar_list) <= 0
                or pillar_list[-1].x < p.get_screen_width() - horitzontal_gap
            ):
                random_height = random.randint(
                    vertical_gap, p.get_screen_height() - vertical_gap
                )

                pillar = p.Rectangle(pillar_x, pillar_y, pillar_width, random_height)
                pillar_list.append(pillar)

            for pillar in pillar_list:
                pillar.x -= pillar_speed

                # update score
                if rect.x == pillar.x + pillar.width:
                    score += 1

                if pillar.x + pillar.width < 0:
                    pillar_list.remove(pillar)

                # pillar collision rect
                if p.check_collision_recs(rect, pillar):
                    game_over = True
                elif p.check_collision_recs(
                    rect,
                    (
                        pillar.x,
                        pillar.height + vertical_gap,
                        pillar.width,
                        p.get_screen_height() - pillar.height - vertical_gap,
                    ),
                ):
                    game_over = True

            if p.is_key_pressed(p.KeyboardKey.KEY_UP):
                change_y = jump_force

            change_y += gravity
            rect.y += change_y

        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        p.draw_rectangle_rec(rect, rect_color)

        for pillar in pillar_list:
            p.draw_rectangle_rec(
                (pillar.x, pillar.y, pillar.width, pillar.height), pillar_color
            )
            p.draw_rectangle_rec(
                (
                    pillar.x,
                    pillar.height + vertical_gap,
                    pillar.width,
                    p.get_screen_height() - pillar.height - vertical_gap,
                ),
                pillar_color,
            )

        # draw score
        p.draw_text(str(score), 20, 20, 40, p.BLACK)

        if game_over:
            p.draw_text(
                "GAME OVER",
                p.get_screen_width() // 2 - 120,
                p.get_screen_height() // 2,
                40,
                p.BLACK,
            )
        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: c_files/raylib/single_file_games/flappy_bird.c
#include "raylib.h"
#include <stdio.h>

#define PILLAR_GAP 200
#define NUMBER_OF_PILLARS 100
#define SCORE_STR_LENGTH 10

// ************************************
// PLAYER
// ************************************
typedef struct Player
{
    float x;
    float y;
    float width;
    float height;
    Color color;
    float jump_force;
    float gravity;
    float change_y;
} Player;

void init_player(Player *player)
{
    player->x = 20;
    player->y = 20;
    player->width = 30;
    player->height = 30;
    player->color = RED;
    player->jump_force = -15;
    player->gravity = 1;
    player->change_y = 0;
}

void draw_player(Player *player)
{
    DrawRectangleRec((Rectangle){player->x,
                                 player->y,
                                 player->width,
                                 player->height},
                     player->color);
}

void update_player(Player *player)
{
    if (IsKeyPressed(KEY_UP))
    {
        player->change_y = player->jump_force;
    }

    player->change_y += player->gravity;
    player->y += player->change_y;
}

// ************************************
// PILLARS
// ************************************
typedef struct Pillar
{
    float x;
    float y;
    float width;
    float height;
    Color color;
    float speed;

} Pillar;

void init_pillar(Pillar *pillar)
{
    pillar->x = GetScreenWidth();
    pillar->y = 0;
    pillar->width = 100;
    pillar->height = GetRandomValue(0, GetScreenHeight() - PILLAR_GAP);
    pillar->color = DARKGREEN;
    pillar->speed = 4;
}

void draw_pillar(Pillar *pillar)
{
    DrawRectangleRec((Rectangle){pillar->x,
                                 pillar->y,
                                 pillar->width,
                                 pillar->height},
                     pillar->color);
    DrawRectangleRec((Rectangle){pillar->x,
                                 pillar->y + pillar->height + PILLAR_GAP,
                                 pillar->width,
                                 GetScreenHeight() - pillar->height},
                     pillar->color);
}

void update_pillar(Pillar *pillar)
{
    pillar->x -= pillar->speed;
}

// ************************************
// PILLAR LIST
// ************************************
typedef struct PillarList
{
    Pillar list[NUMBER_OF_PILLARS];
    int size;

} PillarList;

void init_pillar_list(PillarList *pillar_list)
{
    pillar_list->size = 0;
}

void draw_pillar_list(PillarList *pillar_list)
{
    for (int i = 0; i < pillar_list->size; i++)
    {
        draw_pillar(&pillar_list->list[i]);
    }
}

void update_pillar_list(PillarList *pillar_list)
{
    if (pillar_list->size <= 0 ||
        pillar_list->list[pillar_list->size - 1].x < 0)
    {
        Pillar pillar;
        init_pillar(&pillar);
        pillar_list->list[pillar_list->size] = pillar;
        pillar_list->size++;
    }

    for (int i = 0; i < pillar_list->size; i++)
    {
        update_pillar(&pillar_list->list[i]);

        for (int i = 0; i < pillar_list->size; i++)
        {
            if (pillar_list->list[i].x + pillar_list->list[i].width < 0)
            {
                pillar_list->size--;
                for (int j = i; j < pillar_list->size; j++)
                {
                    pillar_list->list[j] = pillar_list->list[j + 1];
                }
            }
        }
    }
}

// ************************************
// MAIN LOOP
// ************************************
int main(void)
{
    const int screenWidth = 400;
    const int screenHeight = 600;
    const char screenTitle[] = "Flappy Bird";
    const Color screenBackground = SKYBLUE;
    const int gameFps = 60;

    InitWindow(screenWidth, screenHeight, screenTitle);
    SetTargetFPS(gameFps);

    int score = 0;
    char score_str[SCORE_STR_LENGTH];
    bool game_over = false;

    Player player;
    PillarList pillar_list;

    init_player(&player);
    init_pillar_list(&pillar_list);

    while (!WindowShouldClose())
    {
        if (!game_over)
        {
            // convert to string
            snprintf(score_str, SCORE_STR_LENGTH, "%d\n", score);

            update_player(&player);
            update_pillar_list(&pillar_list);

            // player collision pillars
            for (int i = 0; i < pillar_list.size; i++)
            {
                if (CheckCollisionRecs((Rectangle){player.x,
                                                   player.y,
                                                   player.width,
                                                   player.height},
                                       (Rectangle){pillar_list.list[i].x,
                                                   pillar_list.list[i].y,
                                                   pillar_list.list[i].width,
                                                   pillar_list.list[i].height}))
                {
                    game_over = true;
                }
                if (CheckCollisionRecs((Rectangle){player.x,
                                                   player.y,
                                                   player.width,
                                                   player.height},
                                       (Rectangle){pillar_list.list[i].x,
                                                   pillar_list.list[i].y + pillar_list.list[i].height + PILLAR_GAP,
                                                   pillar_list.list[i].width,
                                                   GetScreenHeight() - pillar_list.list[i].height}))
                {
                    game_over = true;
                }

                // update scores
                if (player.x == pillar_list.list[i].x + pillar_list.list[i].width)
                {
                    score++;
                }
            }
        }
        else
        {
            if (IsKeyPressed(KEY_ENTER))
            {
                score = 0;
                game_over = false;
                init_player(&player);
                init_pillar_list(&pillar_list);
            }
        }

        BeginDrawing();
        ClearBackground(screenBackground);

        draw_player(&player);
        draw_pillar_list(&pillar_list);

        // draw score
        DrawText(score_str, GetScreenWidth() - 80, 20, 40, BLACK);

        // game over
        if (game_over)
        {
            DrawText("Game Over", GetScreenWidth() / 2 - 70, GetScreenHeight() / 2, 30, BLACK);
        }

        EndDrawing();
    }

    CloseWindow();

    return 0;
}